Bluetooth: services: Adds additonal len checks for ots_dir_list

Thalley opened the pull request:
Adds additional length checks for the OTS directory listing
implementation. This will check the object name length and
the total length of the object when encoding,
as well as the length of the objects when removing objects
from the directory listing.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/33830

Thalley committed:
Bluetooth: services: Adds additonal len checks for ots_dir_list

Adds additional length checks for the OTS directory listing
implementation. This will check the object name length and
the total length of the object when encoding,
as well as the length of the objects when removing objects
from the directory listing.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -112,10 +124,20 @@ void bt_ots_dir_list_obj_remove(struct bt_ots_dir_list *dir_list, void *obj_mana
+		id = sys_get_le64(dir_list->net_buf.data + offset + sizeof(len));
 
 		__ASSERT(len, "Invalid object length");
+		__ASSERT((len + offset) <= dir_list->net_buf.len,
+			 "Invalid dir_list buf length %u, expected at least %u",
+			 dir_list->net_buf.len, len + offset);
```
Does `dir_list.net_buf` originate from some remote peer device? If so, then assert is the wrong choice here. That should only be used to catch local programming bugs, whereas invalid data from a peer should be handled gracefully.

Thalley replied:
The dir_list.net_buf (in fact everything related to the dir_list) is created by the OTS service locally. It is an internal read-only object. 

The dir_list object is, however, created based on metadata from other, potentially remotely created objects, but my opinion is that it is the responsibility of the OTS implementation (and not the OTS dir list implementation) to validate remotely created objects. That being said, the current version of our OTS implementation does not allow a client to create objects, but only the local application.

carlescufi commented:
@kapi-no can you please review?

kapi-no commented on subsys/bluetooth/services/ots/ots_dir_list.c:
```diff
@@ -41,7 +43,10 @@ static void dir_list_object_encode(struct bt_gatt_ots_object *obj,
-	net_buf_simple_add_u8(net_buf, strlen(obj->metadata.name));
+	obj_name_len = strlen(obj->metadata.name);
+	__ASSERT(obj_name_len >= 0 && obj_name_len <= 120,
+		 "Dir list object len is too small %u", len);
```
```suggestion
		 "Dir list object len is incorrect %u", len);
```

It can be too large.

The first condition seems always true:

```obj_name_len >= 0 ```

Thalley replied:
The first condition should have been `obj_name_len > 0` as empty strings are not allowed for the names.

Thalley committed:
Bluetooth: services: Adds OTS name length checks

Adds name length checks. The OTS spec does not
explicitely specifiy a maximum name length, but the
maximum name length in the directory listing object
shall be less or equal to 120 octets.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@jhedberg Added another commit w.r.t. name length checks. Will you please re-review?
